[Xtensa] Implement lowering SELECT_CC, SETCC.

andreisfr opened the pull request.

andreisfr committed:
[Xtensa] Implement lowering SELECT_CC, SETCC.

andreisfr commented:
@arsenm , @jyknight,  @MaskRay , @s-barannikov , PTAL to this PR whenever you have the time.

andreisfr committed:
[Xtensa] Fix setcc test and remove redundant code.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -90,6 +90,22 @@ XtensaTargetLowering::XtensaTargetLowering(const TargetMachine &TM,
   setOperationAction(ISD::BR_CC, MVT::f32, Expand);
 
+  // Used by legalize types to correctly generate the setcc result.
+  setOperationPromotedToType(ISD::SETCC, MVT::i1, MVT::i32);
+  setOperationPromotedToType(ISD::BR_CC, MVT::i1, MVT::i32);
```
Looks redundant, this should be the default behavior.
Also, BR_CC case is repeated above.

andreisfr replied:
@s-barannikov , thank you very much for comments. I removed this code and add more tests for branches.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -90,6 +90,22 @@ XtensaTargetLowering::XtensaTargetLowering(const TargetMachine &TM,
+  setOperationPromotedToType(ISD::BR_CC, MVT::i1, MVT::i32);
+
+  setOperationAction(ISD::BR_CC, MVT::i32, Legal);
+  setOperationAction(ISD::BR_CC, MVT::i64, Expand);
```
This should happen by default since i64 is illegal. Same for every other node.

andreisfr replied:
Fixed

andreisfr commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -90,6 +90,26 @@ XtensaTargetLowering::XtensaTargetLowering(const TargetMachine &TM,
   setOperationAction(ISD::BR_CC, MVT::f32, Expand);
 
+  // Used by legalize types to correctly generate the setcc result.
+  // AddPromotedToType(ISD::SETCC, MVT::i1, MVT::i32);
```
@arsenm, thank you very much for comments! Fixed

andreisfr commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -90,6 +90,26 @@ XtensaTargetLowering::XtensaTargetLowering(const TargetMachine &TM,
+  setOperationAction(ISD::SETCC, MVT::i32, Custom);
+  setOperationAction(ISD::SETCC, MVT::i64, Expand);
+
+  // make BRCOND legal, its actually only legal for a subset of conds
```
I removed code related to BRCOND from this PR, seems it is redundant , I will add it in next patches.

andreisfr commented on llvm/test/CodeGen/Xtensa/select-cc.ll:
```diff
@@ -0,0 +1,125 @@
+
+  %val27 = load i32, ptr %b
+  %tst14 = icmp ugt i32 %val21, 2046
+  %val28 = select i1 %tst14, i32 %val26, i32 %val27
```
I created set of functions for each compare case instead of single function.

andreisfr commented on llvm/utils/UpdateTestChecks/asm.py:
```diff
@@ -109,6 +109,13 @@ class string:
+    r'(?P<body>.*?)\n'
+    r'^\.Lfunc_end\d+:\n',  # Match the end label
+    flags=(re.M | re.S)
+)
```
I removed asm.py modification from this PR

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -514,6 +530,38 @@ XtensaTargetLowering::LowerReturn(SDValue Chain, CallingConv::ID CallConv,
+  ISD::CondCode CC = cast<CondCodeSDNode>(Op.getOperand(2))->get();
+  SDValue TargetCC = DAG.getConstant(CC, DL, MVT::i32);
+
+  // Expand to target SELECT_CC
```
Isn't this what `Expand` operation action does?

andreisfr replied:
Yes, at current stage it produce the same result, so I removed custom setcc lowering, I will add specific version in next patches. Fixed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaOperators.td:
```diff
@@ -38,3 +43,7 @@ def Xtensa_callseq_end  : SDNode<"ISD::CALLSEQ_END",   SDT_XtensaCallSeqEnd,
 
 def Xtensa_brjt: SDNode<"XtensaISD::BR_JT", SDT_XtensaBrJT, [SDNPHasChain]>;
+
+def Xtensa_select : SDNode<"XtensaISD::SELECT", SDTSelect>;
```
Unused node

andreisfr replied:
Removed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.h:
```diff
@@ -33,7 +33,14 @@ enum {
+  // mask of condition-code values for which operand 0 should be
+  // chosen over operand 1; it has the same form as BR_CCMASK.
+  // Operand 3 is the flag operand.
+  SELECT,
```
SELECT is unused

andreisfr replied:
Removed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -697,6 +749,136 @@ const char *XtensaTargetLowering::getTargetNodeName(unsigned Opcode) const {
+static int GetBranchKind(int Cond, bool &BrInv) {
+  switch (Cond) {
+  case ISD::SETEQ:
+  case ISD::SETOEQ:
+  case ISD::SETUEQ:
```
Floating-point comparisons shouldn't reach here as FP types are illegal.

andreisfr replied:
Fixed

s-barannikov reviewed (commented):
Need tests for branches

andreisfr committed:
[Xtensa] Minor fixes, remove redundant code.

Removed redundant operations promotions and actions. Also removed setcc
lowering and redundant branch kind cases from GetBranchKind() function.
Implemented additional br_cc tests.

s-barannikov commented on llvm/test/CodeGen/Xtensa/setcc.ll:
```diff
@@ -0,0 +1,231 @@
+; RUN: llc < %s -mtriple=xtensa-unkonwn-elf -O0 | FileCheck %s
```
All llc tests should have -verify-machineinstrs argument.

s-barannikov commented on llvm/test/CodeGen/Xtensa/select-cc.ll:
```diff
@@ -0,0 +1,184 @@
+  %tst1 = icmp ult i32 %a, 1024
+  %val2 = select i1 %tst1, i32 %a, i32 %val1
+  ret i32 %val2
+}
```
Please add tests with i64 icmp and i64 select.

andreisfr replied:
I added tests for SELECTCC/SETCC with i64 operands.

s-barannikov commented on llvm/test/CodeGen/Xtensa/select-cc.ll:
```diff
@@ -0,0 +1,184 @@
+; RUN: llc -mtriple=xtensa -disable-block-placement -verify-machineinstrs < %s \
+; RUN:   | FileCheck %s
+
+define signext i32 @f_eq(i32 signext %a, ptr %b) nounwind {
```
signext is redundant

s-barannikov commented on llvm/test/CodeGen/Xtensa/brcc.ll:
```diff
@@ -1,10 +1,10 @@
 ; RUN: llc -march=xtensa < %s | FileCheck %s
 
-; CHECK-LABEL: brcc1:
+; CHECK-LABEL: brcc_sgt:
```
Autogenerate check lines?

s-barannikov replied:
It isn't, and -verify-machineinstrs is also missing

arsenm replied:
-verify-machineinstrs should be missing. We should stop unconditionally adding it to every test, and only in cases specifically testing things that hit a verifier error

s-barannikov replied:
It should never be hit, unless it is an XFAIL test.
From my experience, adding -verify-machineinstrs to llc tests saves a lot of time catching bugs that otherwise could be caught only when building large applications.
Machine verifier is effectively `assert(mir.isValid())`, I don't see a reason why you woulndn't want this assert when running tests.

I don't insist though (I would if I were a maintainer of this backend).

andreisfr replied:
I added "-verify-machineinstrs" argument to the brcc.ll test and generated check lines.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -697,6 +718,126 @@ const char *XtensaTargetLowering::getTargetNodeName(unsigned Opcode) const {
+  case ISD::SETULT:
+    return Xtensa::BLTU;
+  case ISD::SETULE:
+    BrInv = true;
```
Illegal condition codes should be handled by `setCondCodeAction(CC, Expand)`.

andreisfr replied:
I'm sorry, you mean to add setCondCodeAction() for codes which are not supported for i32 type natively by Xtensa? Currently all basic integer condition codes are supported, only codes like SETOLE which are used with FP types are not handled. Maybe I missed something.

s-barannikov replied:
If I'm reading `XtensaInstrInfo.td` correctly, there are no instructions corresponding to ULE and UGT condition codes. These codes should be legalized by `setCondCodeAction(..., Expand)` and removed from this switch.

andreisfr replied:
I added patterns for brcc with  ULE and UGT codes:
[ brcc operation](https://github.com/llvm/llvm-project/blob/ac634c8b08adef1da8114821fbb7ad8a70e994a2/llvm/lib/Target/Xtensa/XtensaInstrInfo.td#L419)
And the same approach is used in lowering SELECTCC. 
So, your suggestion is to remove such ULE/UGT handling in code and use setCondCodeAction(CC, Expand) instead?

s-barannikov replied:
> So, your suggestion is to remove such ULE/UGT handling in code and use setCondCodeAction(CC, Expand) instead?

Yes. That would allow you to remove those patterns and BrInv.

andreisfr replied:
It's interesting that if I use setCondCodeAction(CC, Expand) code instead of patterns then llc generates a bit large code in some cases, for example I compiled following code with "O3" option:
`
define i32 @select_ugt(i32 %a, ptr %b) nounwind {
  %val1 = load i32, ptr %b
  %tst1 = icmp ugt i32 %a, %val1
  %val2 = select i1 %tst1, i32 %a, i32 %val1
  ret i32 %val2
}

Old llc version:

select_ugt:
	l32i	a8, a3, 0
	bltu	a8, a2, .LBB1_2
	or	a2, a8, a8
.LBB1_2:
	ret

New llc version:

select_ugt:
	or	a8, a2, a2
	l32i	a2, a3, 0
	bgeu	a2, a8, .LBB1_2
	or	a2, a8, a8
.LBB1_2:
	ret
`

Should we strive to always rely on general approaches to lower Operations(like use setCondCodeAction), or in some cases (like that) we can use target specific implementation?

andreisfr replied:
I implemented expand action for GT/UGT/LE/ULE code conditions and remove patterns.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -514,6 +517,22 @@ XtensaTargetLowering::LowerReturn(SDValue Chain, CallingConv::ID CallConv,
+  SDValue TrueValue = Op.getOperand(2);
+  SDValue FalseValue = Op.getOperand(3);
+  ISD::CondCode CC = cast<CondCodeSDNode>(Op->getOperand(4))->get();
+  SDValue TargetCC = DAG.getConstant(CC, DL, MVT::i32);
```
Condition code should be converted to target opcode here rather than in emitSelectCC.

andreisfr committed:
[Xtensa] Add tests for SELECTCC/SETCC with i64 operands.

Implemented new tests for SELECTCC/SETCC with i64 operands. Added
minor code fixes.

andreisfr commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -697,6 +718,126 @@ const char *XtensaTargetLowering::getTargetNodeName(unsigned Opcode) const {
+  case ISD::SETUGE:
+    return Xtensa::BGEU;
+  default:
+    return -1;
```
Fixed

andreisfr commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -697,6 +718,126 @@ const char *XtensaTargetLowering::getTargetNodeName(unsigned Opcode) const {
+// Custom insertion
+//===----------------------------------------------------------------------===//
+
+static int GetBranchKind(int Cond, bool &BrInv) {
```
Corrected

andreisfr commented on llvm/lib/Target/Xtensa/XtensaInstrInfo.td:
```diff
@@ -574,3 +575,12 @@ let Defs = [SP], Uses = [SP] in {
+let usesCustomInserter = 1 in {
+  def SELECT : Pseudo<(outs AR:$dst), (ins AR:$lhs, AR:$rhs, AR:$t, AR:$f, i32imm:$cond),
+                     "!select $dst, $lhs, $rhs, $t, $f, $cond",
+                     [(set AR:$dst, (Xtensa_select_cc AR:$lhs, AR:$rhs, AR:$t, AR:$f, imm:$cond))]>;
```
Fixed

andreisfr commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -697,6 +718,126 @@ const char *XtensaTargetLowering::getTargetNodeName(unsigned Opcode) const {
+  MBB->addSuccessor(CopyMBB);
+  MBB->addSuccessor(SinkMBB);
+
+  bool BrInv = false;
```
Fixed

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -514,6 +517,62 @@ XtensaTargetLowering::LowerReturn(SDValue Chain, CallingConv::ID CallConv,
+  if (BrInv) {
+    return DAG.getNode(XtensaISD::SELECT_CC, DL, Ty, RHS, LHS, TrueValue,
+                       FalseValue, TargetCC);
+  } else {
```
No else after return. Also could just if (BrInv) swap values?

andreisfr replied:
Fixed

andreisfr committed:
[Xtensa] Minor fixes in code.

andreisfr committed:
[Xtensa] Expand actions for GT/UGT/LE/ULE code condtions.

The GT/UGT/LE/ULE code condtions not supported for i32, some expand actions for these codes.
Also corrected brcc.ll/select-cc.ll/setcc.ll tests.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -514,6 +522,51 @@ XtensaTargetLowering::LowerReturn(SDValue Chain, CallingConv::ID CallConv,
+  SDValue FalseValue = Op.getOperand(3);
+  ISD::CondCode CC = cast<CondCodeSDNode>(Op->getOperand(4))->get();
+
+  unsigned BrKind = getBranchOpcode(CC);
```
(nit) BrOpcode

andreisfr replied:
Fixed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -514,6 +522,51 @@ XtensaTargetLowering::LowerReturn(SDValue Chain, CallingConv::ID CallConv,
+  unsigned BrKind = getBranchOpcode(CC);
+  SDValue TargetCC = DAG.getConstant(BrKind, DL, MVT::i32);
+
+  // Wrap select nodes
```
It doesn't wrap anything

andreisfr replied:
Fixed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -697,6 +752,86 @@ const char *XtensaTargetLowering::getTargetNodeName(unsigned Opcode) const {
+
+  MachineFunction *F = MBB->getParent();
+  MachineBasicBlock *CopyMBB = F->CreateMachineBasicBlock(LLVM_BB);
+  MachineBasicBlock *SinkMBB = F->CreateMachineBasicBlock(LLVM_BB);
```
You will need a code like this here (for a future patch that adds related tests), see 2dcf0512:
```
  // Set the call frame size on entry to the new basic blocks.
  uint64_t CallFrameSize = TII.getCallFrameSizeAt(MI);
  CopyMBB->setCallFrameSize(CallFrameSize);
  SinkMBB->setCallFrameSize(CallFrameSize);
```

andreisfr replied:
Thank you, I will use this information in future.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.h:
```diff
@@ -33,7 +33,13 @@ enum {
+  // a false value (ops #2 and #3) based on the boolean result of comparing
+  // the lhs and rhs (ops #0 and #1) of a conditional expression with the
+  // condition code in op #4
+  SELECT_CC
```
```suggestion
  SELECT_CC,
```

andreisfr replied:
Fixed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.h:
```diff
@@ -95,6 +114,10 @@ class XtensaTargetLowering : public TargetLowering {
 
   CCAssignFn *CCAssignFnForCall(CallingConv::ID CC, bool IsVarArg) const;
+
+  // Implement EmitInstrWithCustomInserter for individual operation types.
```
The comment isn't helpful.

andreisfr replied:
Fixed

s-barannikov reviewed (approved):
Some final nits

andreisfr committed:
[Xtensa] Minor fixes.

Renamed variable and removed redundant comments.
